### DISCUSSION TITLE
Article feedback

### DISCUSSION THREAD

**Discussion by scasagrande** [scasagrande/toolchains_llvm_sysroot#1] [General]
If you have any feedback for my corresponding article be sure to leave it here!

https://steven.casagrande.io/articles/sysroot-generation-toolchains-llvm

**Comment by irajtaghlidi** [comment]
Nice article @scasagrande,
I am trying to create a fully hermetic CC toolchain for my Rust project in Bazel in RBE mode.
After fixing many different problems I reached a point where I needed to provide some shared libraries that only exist in my custom llvm `sysroot` to `rules_rust` to build an `*-sys` crate which `build.rs` even depends on `CMake`. But I still don't know what is the right direction to focus on. As the last line of your blog post mentioned, I'm so curious to read more about your experience with this part.

This is snap of my setup:

```
# WORKSPACE

# sysroot
#############################
http_archive(
    name = "org_chromium_sysroot_linux_x64",
    build_file_content = """
filegroup(
  name = "sysroot",
  srcs = glob(["*/**"]),
  visibility = ["//visibility:public"],
)
""",
    sha256 = "484b6082237a466219fa960b360c3acc4ace5b57941d422f79a2768843823061",
    urls = ["http://localhost:8000//debian_bookworm_amd64_sysroot.tar.xz"],
)


# LLVM toolchain with sysroot
#############################
...
llvm_toolchain(
    name = "llvm_toolchain_linux_exec",
    exec_arch = "amd64",
    exec_os = "linux",
    llvm_version = "12.0.0",
    sysroot = {
        "linux-x86_64": "@org_chromium_sysroot_linux_x64//:sysroot",
    },
    strip_prefix = {"": "clang+llvm-12.0.0-x86_64-linux-gnu-ubuntu-20.04"},
    urls = {"": ["https://github.com/llvm/llvm-project/releases/download/llvmorg-12.0.0/clang+llvm-12.0.0-x86_64-linux-gnu-ubuntu-20.04.tar.xz"]},
)

# ForeignCc
#############################
rules_foreign_cc_dependencies(
    native_tools_toolchains = [],
    register_built_tools = False,  # do not build cmake/make from source,
    register_preinstalled_tools = False,  # nor use pre-installed.
    register_default_tools = True,  # just download from official site,
    cmake_version="3.25.1",
    register_built_pkgconfig_toolchain = True
)


# rules_rust
#############################
...
load("@rules_rust//crate_universe:defs.bzl", "crate", "crates_repository", "splicing_config")
crates_repository(
    name = "crate_index",                # Name of the repository to generate.
    cargo_lockfile = "//:Cargo.lock", 
    lockfile = "//:Cargo.Bazel.lock",
    manifests = [                        # List of Cargo.toml files from which to
        "//:Cargo.toml",                 # build the dependency graph. The manifests
        "//:service1/Cargo.toml",        # of all crates in the workspace must be
        "//:common/lib1/Cargo.toml",     # listed here.
        "//:common/lib2/Cargo.toml",
        "//:common/lib3-using-mbedtls/Cargo.toml",
    ],
    annotations = {
        "mbedtls-sys-auto": [crate.annotation(
            # Setting build_script_data makes the files available on disk when the rule runs.
            build_script_data = ["@rules_foreign_cc//toolchains:current_cmake_toolchain", "@org_chromium_sysroot_linux_x64//:sysroot"],
            build_script_env = {
                # The toolchain supplies a value of $(CMAKE) which is an execroot-relative path, so we need to prefix it with $${pwd}/ because build scripts don't typically run in the execroot unlike most bazel rules, for improved compatibility with Cargo.
                "CMAKE": "$${pwd}/$(CMAKE)",
                # "LIBCLANG_PATH": "$(location @org_chromium_sysroot_linux_x64//lib)",
            },
            # Setting build_script_toolchains makes makefile variable substitution work so that we can reference $(CNAME) in attributes.
            build_script_toolchains = ["@rules_foreign_cc//toolchains:current_cmake_toolchain"],
        )],
        "mbedtls": [crate.annotation(
            build_script_data = ["@crate_index__mbedtls-sys-auto-2.28.7//:mbedtls_headers"],
            build_script_env = {
                "DEP_MBEDTLS_INCLUDE_BZL": "$(location @crate_index__mbedtls-sys-auto-2.28.7//:mbedtls_headers)",
            },
        )],
    },
    rust_version = "nightly/2023-05-07", 
    splicing_config = splicing_config(
        resolver_version = "2",
    ),
)

```


but I still face this error, and trying to figure out what is the best solution to fix LIBCLANG_PATH issue or use a single sysroot for CC and Rust.
```
thread 'main' panicked at 'Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"', external/crate_index__bindgen-0.65.1/lib.rs:603:31
stack backtrace:
   0:     0x564a6be84231 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbf761e313ca3ada9
   1:     0x564a6bead69f - core::fmt::write::h66441248ae33813a
```

I would appreciate if you can take a look and give some advices.

Here is a remotee execution log:
https://app.buildbuddy.io/invocation/404ffa75-e609-48ca-ba46-8c2c691019d9

**Reply by scasagrande** [reply]
my suggestion is to avoid calling the `build.rs` build processes as much as possible. It quickly becomes a mess if you have bazel calling rust calling build.rs calling cmake :)

I suggest that you instead replicate the `*-sys` build processes in native bazel, and then bypass the build script entirely

for example, lets say you have a rust crate dependency on https://crates.io/crates/rdkafka

this crate in turn has a dependency on https://crates.io/crates/rdkafka-sys/ . This crate's job is basically just to compile librdkafka

instead what I do is build librdkafka in bazel, and then provide this bazel target label as a to the crate annotation

for example, lets assume I have a bazel build for librdkafka available at `@librdkafka//:librdkafka`. My crate annotation would look like the following:

```
"rdkafka-sys": [crate.annotation(
    gen_build_script = False,
    deps = ["@librdkafka//:librdkafka"],
)],
```

**Reply by irajtaghlidi** [reply]
Thanks for your reply.
The only thing that stopped me from choosing the path you suggested was the size of the Rust build script.
Do you think it is possible to rewrite such a big build script in native Bazel with CMake and bindgen targets?
https://github.com/fortanix/rust-mbedtls/tree/master/mbedtls-sys/build

What if it would not be possible? What would be the second option for us? Assuming we only face one Rust sys crate with CMake as dependency and can use native rules for others.

**Reply by scasagrande** [reply]
another option is to pre-build the artifacts (`.so` etc) using plain `cargo build` for each of your different supported platforms, and then feed them into the crate annotation in the same way

for example, I want to use a pre-built copy of openssl:

```
"openssl-sys": [crate.annotation(
    gen_build_script = False,
    deps = ["@mono//bazel/external/openssl"],
)],
```
where `@mono` is the name of our current bazel workspace

then in `//bazel/external/openssl`

```
alias(
    name = "openssl",
    actual = select({
        "@bazel_tools//src/conditions:darwin_arm64": "@openssl_macos_arm64//:openssl",
        "@bazel_tools//src/conditions:darwin_x86_64": "@openssl_macos_x86_64//:openssl",
        "@bazel_tools//src/conditions:linux_aarch64": "@sysroot_linux_aarch64_2_28//:libssl",
        "@bazel_tools//src/conditions:linux_x86_64": "@sysroot_linux_x86_64_2_28//:libssl",
    }),
    visibility = ["//visibility:public"],
)
```

where this will select the correct pre-built copy of openssl

**Reply by irajtaghlidi** [reply]
This is a good idea, I didn't know it was possible. Thanks for the hints.
But still, not an ideal solution for my situation since we decided to don't use Cargo at all. Needs to figure out how I can add a new LIB_PATH to the task.

**Reply by scasagrande** [reply]
I don't use cargo either, I'm just suggesting that you clone the repo in question and run its build separately. Then package up the files into `.tar.gz` archives, and upload them to whatever artifact store you are using

**Reply by irajtaghlidi** [reply]
I got it, will try to test it then.
Thanks for all your help.

**Reply by irajtaghlidi** [reply]
Hi Again,
before any major change in setup, I found an issue with my sysroot which was a missing symlink for `libclang.so`. But even after setting the `LIBCLANG_PATH` to the correct location I facing this error:
* All mentioned shared libraries exist.
* 

```
thread 'main' panicked at 'Unable to find libclang: "the `libclang` shared library at /buildbuddy/remotebuilds/13c3ecba-9375-453f-8ab5-09e7fb8d572a/external/org_chromium_sysroot_linux_x64/usr/lib/x86_64-linux-gnu/libclang-15.so.15.0.6 could not be opened: libLLVM-15.so.1: cannot open shared object file: No such file or directory"', external/crate_index__bindgen-0.65.1/lib.rs:603:31
stack backtrace:
   0:     0x5564abee2d71 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbf761e313ca3ada9
```

The whole sysroot also is shared with all related crates and SHOULD have access. Do you think is still a sandboxing issue or something else?
```
"mbedtls-sys-auto": [crate.annotation(
          build_script_data = ["@org_chromium_sysroot_linux_x64//:sysroot"],
          build_script_env = {
              "LIBCLANG_PATH": "$${pwd}/external/org_chromium_sysroot_linux_x64/usr/lib/x86_64-linux-gnu",
          },
      )],
```

https://app.buildbuddy.io/invocation/545579aa-31ea-4958-bb57-e08caa50db22

**Reply by scasagrande** [reply]
again I do not try to run rust builds scripts that in turn run cmake. The underlying cmake call is probably not hermetic at all and may not even be using your sysroot

**Comment by jbrindza** [comment]
Hi, Thank you for providing the tutorial and examples. I am attempting to use the sysroot. I can build it and have it integrated into my bazel build, however when I try link I get this error (I get the same error building the sysroot myself and with the one you released):

```
ld.lld: error: external/sysroot_linux_x86_64+/usr/lib/crtn.o:(.debug_line) is compressed with ELFCOMPRESS_ZLIB, but lld is not built with zlib support
```

I am using the llvm v16.0.0 from llvm_toolchain. Do you have any pointers on how to fix this? Thank you.

**Reply by scasagrande** [reply]
that would be an issue with the version of llvm that you are using as it was built without zlib support. You can try 15.0.2 to validate your setup, but then you'll want to build your own copies of llvm that suite your needs.

alternatively you can try the llvm copies uploaded here, but note that they were built against a fairly recent version of glibc and may not work on your system, depending on your host OS. https://github.com/MaterializeInc/toolchains/releases

**Reply by jbrindza** [reply]
Thanks! I just tried 15.0.2 and that worked great. I don't have an particular need for 16.0.0 so I can just stick with this. thank you

**Comment by kdowney-lot49** [comment]
Thanks for the article! I extended this idea to create a fully-dynamic sysroot builder:

https://paragraph.com/@kyle-downey/cross-compiling-c-with-bazel-and-wolfi
